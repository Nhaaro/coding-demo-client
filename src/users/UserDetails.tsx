import {
  Box,
  Breadcrumbs,
  Button,
  Card,
  CardContent,
  Grid,
  Link,
  Stack,
  Toolbar,
  Typography,
} from '@mui/material'
import { useContext } from 'react'
import { useTranslation } from 'react-i18next'
import { Link as RouterLink, useParams } from 'react-router-dom'
import { UserContext } from './UserContext'

const UserDetails = () => {
  const { t } = useTranslation()
  const { userId } = useParams()
  const userContext = useContext(UserContext)

  const user = userContext?.users.find(
    (user) => user.id == parseInt(userId || '')
  )

  return !user ? (
    <div>no user</div>
  ) : (
    <>
      <Toolbar>
        <Box sx={{ flexGrow: 1 }}>
          <Breadcrumbs aria-label="breadcrumb">
            <Link
              underline="hover"
              color="inherit"
              component={RouterLink}
              to="/users"
            >
              <Typography
                variant="h6"
                sx={{
                  fontFamily: 'BlogScript',
                  fontWeight: 700,
                  lineHeight: '2rem', // prevent letters being cropped
                  mb: '-0.3rem', // align with breadcrumbs
                }}
                color="secondary"
              >
                {t('Navigation.Tabs.Users')}
              </Typography>
            </Link>
            <Typography color="text.primary">{userId}</Typography>
          </Breadcrumbs>
        </Box>
        <Stack spacing={2} direction="row" sx={{ flexGrow: 0 }}>
          <Button
            variant="outlined"
            component={RouterLink}
            to={`/users/${userId}/edit`}
            role="button"
          >
            {t('Users.Actions.UpdateUser')}
          </Button>
        </Stack>
      </Toolbar>
      <Grid container spacing={1}>
        <Grid item xs={12} sm={12} md={4}>
          <Card>
            <CardContent
              sx={{
                display: 'flex',
                flexDirection: 'column',
                '& img': {
                  borderRadius: '50%',
                  padding: '3rem',
                },
              }}
            >
              <img
                src={`https://i.pravatar.cc/150?img=${userId}`}
                alt={`https://i.pravatar.cc/150?img=${userId}`}
              />
              <Typography textAlign={'center'} color={'primary'}>
                @{user.username}
              </Typography>
              <Typography textAlign={'center'} variant="h4">
                {user.firstName + ' ' + user.lastName}
              </Typography>
            </CardContent>
          </Card>
        </Grid>
        <Grid item xs={12} sm={12} md={8}>
          <Stack spacing={1}>
            <Card>
              <CardContent>
                <Typography
                  variant="h6"
                  sx={{
                    fontFamily: 'BlogScript',
                    fontWeight: 700,
                    lineHeight: '2rem', // prevent letters being cropped
                    mb: '-0.3rem', // align with breadcrumbs
                  }}
                  color="secondary"
                >
                  {t('Users.Details.AccountDetails').toUpperCase()}
                </Typography>
                <Box sx={{ display: 'flex' }}>
                  <Box sx={{ flex: 1 }}>
                    <Typography>
                      <strong>{t('Users.Details.ID')}</strong>
                      {user.id}
                    </Typography>
                    <Typography>
                      <strong>{t('Users.Details.Status.Label')}</strong>
                      {user.active
                        ? t('Users.Details.Status.Active')
                        : t('Users.Details.Status.Inactive')}
                    </Typography>
                  </Box>
                  <Box sx={{ flex: 1 }}>
                    <Typography>
                      <strong>{t('Users.Details.Email')}</strong>
                      {user.email}
                    </Typography>
                    <Typography>
                      <strong>{t('Users.Details.CreatedAt')}</strong>
                      {new Date(user.createdAt).toLocaleDateString()}
                    </Typography>
                  </Box>
                </Box>
              </CardContent>
            </Card>
            <Card>
              <CardContent>
                <Typography
                  variant="h6"
                  sx={{
                    fontFamily: 'BlogScript',
                    fontWeight: 700,
                    lineHeight: '2rem', // prevent letters being cropped
                    mb: '-0.3rem', // align with breadcrumbs
                  }}
                  color="secondary"
                >
                  {t('Users.Details.PersonalDetails').toUpperCase()}
                </Typography>
                <Box sx={{ display: 'flex' }}>
                  <Box sx={{ flex: 1 }}>
                    <Typography>
                      <strong>{t('Users.Details.FirstName')}</strong>
                      {user.firstName}
                    </Typography>
                  </Box>
                  <Box sx={{ flex: 1 }}>
                    <Typography>
                      <strong>{t('Users.Details.LastName')}</strong>
                      {user.lastName}
                    </Typography>
                  </Box>
                </Box>
              </CardContent>
            </Card>
          </Stack>
        </Grid>
      </Grid>
    </>
  )
}

export default UserDetails
