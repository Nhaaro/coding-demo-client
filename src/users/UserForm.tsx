import SaveIcon from '@mui/icons-material/Save'
import {
  Box,
  Breadcrumbs,
  Button,
  Card,
  CardContent,
  Link,
  Stack,
  TextField,
  Toolbar,
  Typography,
} from '@mui/material'
import React, { FormEvent, useContext, useEffect, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { Link as RouterLink, useParams } from 'react-router-dom'
import { User, UserContext } from './UserContext'

const initialFormData = {
  username: '',
  firstName: '',
  lastName: '',
  email: '',
} as User

const UserForm = () => {
  const { t } = useTranslation()
  const { userId } = useParams()
  const userContext = useContext(UserContext)

  const user = userContext?.users.find(
    (user) => user.id == parseInt(userId || '')
  )

  const [formData, setFormData] = useState<User>(initialFormData)
  useEffect(() => {
    setFormData({
      ...initialFormData,
      ...user,
    })
  }, [user])

  const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setFormData({ ...formData, [event.target.name]: event.target.value })
  }
  const handleSubmit = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault()
  }

  return (
    <form onSubmit={handleSubmit} role="form">
      <Toolbar>
        <Box sx={{ flexGrow: 1 }}>
          <Breadcrumbs aria-label="breadcrumb">
            <Link
              underline="hover"
              color="inherit"
              component={RouterLink}
              to="/users"
            >
              <Typography
                variant="h6"
                sx={{
                  fontFamily: 'BlogScript',
                  fontWeight: 700,
                  lineHeight: '2rem', // prevent letters being cropped
                  mb: '-0.3rem', // align with breadcrumbs
                }}
                color="secondary"
              >
                {t('Navigation.Tabs.Users')}
              </Typography>
            </Link>
            {user ? (
              [
                <Link
                  underline="hover"
                  color="inherit"
                  component={RouterLink}
                  to={`/users/${userId}`}
                  key={1}
                >
                  <Typography>{userId}</Typography>
                </Link>,
                <Typography color="text.primary" key={2}>
                  {t('Navigation.Actions.UpdateUser')}
                </Typography>,
              ]
            ) : (
              <Typography color="text.primary">
                {t('Navigation.Actions.CreateUser')}
              </Typography>
            )}
          </Breadcrumbs>
        </Box>
        <Stack spacing={2} direction="row" sx={{ flexGrow: 0 }}>
          <Button
            variant="outlined"
            component={RouterLink}
            to={user ? `/users/${userId}` : '/users'}
          >
            {t('Users.Actions.Cancel')}
          </Button>
          <Button variant="contained" endIcon={<SaveIcon />} type="submit">
            {t('Users.Actions.SaveUser')}
          </Button>
        </Stack>
      </Toolbar>
      <Stack spacing={1}>
        <Card>
          <CardContent>
            <Typography
              variant="h6"
              sx={{
                fontFamily: 'BlogScript',
                fontWeight: 700,
                lineHeight: '2rem', // prevent letters being cropped
                mb: '-0.3rem', // align with breadcrumbs
              }}
              color="secondary"
            >
              {t('Users.Details.AccountDetails').toUpperCase()}
            </Typography>
            <Stack spacing={1} direction={'row'}>
              <Stack spacing={1} sx={{ flex: 1 }}>
                <TextField
                  fullWidth
                  id="username"
                  name="username"
                  label={t('Users.Attributes.Username')}
                  value={formData.username}
                  onChange={handleInputChange}
                  disabled={!!user}
                />
              </Stack>
              <Stack spacing={1} sx={{ flex: 1 }}>
                <TextField
                  fullWidth
                  id="email"
                  name="email"
                  label={t('Users.Attributes.Email')}
                  value={formData.email}
                  onChange={handleInputChange}
                />
              </Stack>
            </Stack>
          </CardContent>
        </Card>
        <Card>
          <CardContent>
            <Typography
              variant="h6"
              sx={{
                fontFamily: 'BlogScript',
                fontWeight: 700,
                lineHeight: '2rem', // prevent letters being cropped
                mb: '-0.3rem', // align with breadcrumbs
              }}
              color="secondary"
            >
              {t('Users.Details.PersonalDetails').toUpperCase()}
            </Typography>
            <Stack spacing={1} direction={'row'}>
              <Stack spacing={1} sx={{ flex: 1 }}>
                <TextField
                  fullWidth
                  id="first-name"
                  name="firstName"
                  label={t('Users.Attributes.FirstName')}
                  value={formData.firstName}
                  onChange={handleInputChange}
                />
              </Stack>
              <Stack spacing={1} sx={{ flex: 1 }}>
                <TextField
                  fullWidth
                  id="last-name"
                  name="lastName"
                  label={t('Users.Attributes.LastName')}
                  value={formData.lastName}
                  onChange={handleInputChange}
                />
              </Stack>
            </Stack>
          </CardContent>
        </Card>
      </Stack>
    </form>
  )
}

export default UserForm
