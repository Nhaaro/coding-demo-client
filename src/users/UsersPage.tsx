import AddIcon from '@mui/icons-material/Add'
import DeleteIcon from '@mui/icons-material/Delete'
import EditIcon from '@mui/icons-material/Edit'
import {
  Box,
  Button,
  IconButton,
  Link as MuiLink,
  Stack,
  Toolbar,
  Tooltip,
  Typography,
} from '@mui/material'
import {
  DataGrid,
  GridActionsCellItem,
  GridColDef,
  GridRowParams,
  GridRowSelectionModel,
  GridRowsProp,
} from '@mui/x-data-grid'
import { useContext, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { Link } from 'react-router-dom'
import { UserContext } from './UserContext'

const UsersPage = () => {
  const { t } = useTranslation()

  const users = useContext(UserContext)

  const columns: GridColDef[] = [
    { field: 'id', headerName: t('Users.Attributes.ID') },
    {
      field: 'username',
      headerName: t('Users.Attributes.Username'),
      minWidth: 180,
      renderCell(params) {
        return (
          <MuiLink
            component={Link}
            to={`/users/${params.id}`}
            underline="hover"
          >
            {params.value}
          </MuiLink>
        )
      },
    },
    {
      field: 'name',
      headerName: t('Users.Attributes.DisplayName'),
      valueGetter: (_, row) => {
        return `${row.firstName || ''} ${row.lastName || ''}`
      },
      minWidth: 250,
    },
    {
      field: 'email',
      headerName: t('Users.Attributes.Email'),
      minWidth: 150,
    },
    {
      field: 'active',
      headerName: t('Users.Attributes.Status'),
      type: 'boolean',
    },
    {
      field: 'createdAt',
      headerName: t('Users.Attributes.CreatedAt'),
      type: 'dateTime',
      width: 200,
    },
    {
      field: 'actions',
      type: 'actions',
      getActions: (params: GridRowParams) => [
        <Tooltip title={t('Users.Actions.UpdateUser')}>
          <IconButton
            aria-label={t('Users.Actions.UpdateUser')}
            component={Link}
            to={`/users/${params.id}/edit`}
            role="menuitem"
          >
            <EditIcon />
          </IconButton>
        </Tooltip>,
        <Tooltip title={t('Users.Actions.DeleteUser')}>
          <GridActionsCellItem
            icon={<DeleteIcon />}
            label={t('Users.Actions.DeleteUser')}
          />
        </Tooltip>,
      ],
    },
  ]
  const rows: GridRowsProp = users?.users || []

  const [rowSelectionModel, setRowsSelectectionModel] =
    useState<GridRowSelectionModel>([])

  return (
    <>
      <Toolbar>
        <Box sx={{ flexGrow: 1 }}>
          <Typography
            variant="h6"
            sx={{
              fontFamily: 'BlogScript',
              fontWeight: 700,
              lineHeight: '2rem', // prevent letters being cropped
              mb: '-0.3rem', // align with breadcrumbs
            }}
            color="secondary"
          >
            {t('Navigation.Tabs.Users')}
          </Typography>
        </Box>
        <Stack spacing={2} direction="row" sx={{ flexGrow: 0 }}>
          <Button variant="outlined" disabled={!rowSelectionModel.length}>
            {t('Users.Actions.DeleteUserBulk')}
          </Button>
          <Button
            variant="contained"
            component={Link}
            to="/users/create"
            endIcon={<AddIcon />}
          >
            {t('Users.Actions.CreateUser')}
          </Button>
        </Stack>
      </Toolbar>
      <DataGrid
        rows={rows}
        columns={columns}
        initialState={{
          pagination: {
            paginationModel: { page: 0, pageSize: 20 },
          },
        }}
        pageSizeOptions={[20, 50, 100]}
        checkboxSelection
        disableRowSelectionOnClick
        autoHeight
        onRowSelectionModelChange={(rowSelectionModel) => {
          setRowsSelectectionModel(rowSelectionModel)
        }}
      />
    </>
  )
}

export default UsersPage
