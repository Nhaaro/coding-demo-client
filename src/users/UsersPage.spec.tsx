import { expect, test } from '@jest/globals'
import { getByRole, render, screen } from '@testing-library/react'
import {
  createMemoryRouter,
  MemoryRouter,
  RouterProvider,
} from 'react-router-dom'
import UserProvider from './UserContext.tsx'
import UsersPage from './UsersPage.tsx'
import userEvent from '@testing-library/user-event'
import UserForm from './UserForm.tsx'
import UserDetails from './UserDetails.tsx'

describe('UsersPage', () => {
  describe('presentation', () => {
    test('it renders', () => {
      render(
        <UserProvider>
          <MemoryRouter>
            <UsersPage />
          </MemoryRouter>
        </UserProvider>
      )
    })

    test('bulk delete is disabled by default', async () => {
      render(
        <UserProvider>
          <MemoryRouter>
            <UsersPage />
          </MemoryRouter>
        </UserProvider>
      )

      const deleteUsersButton = screen.getByRole('button', {
        name: /Users.Actions.DeleteUserBulk/i,
      })

      expect(deleteUsersButton).toBeDisabled()
    })

    test('table starts with correct data', async () => {
      render(
        <UserProvider>
          <MemoryRouter>
            <UsersPage />
          </MemoryRouter>
        </UserProvider>
      )

      const rows = screen.getAllByRole('row', { name: /Select row/i })

      expect(rows).toMatchSnapshot()
    })

    test('bulk delete button activates on row selection', async () => {
      render(
        <UserProvider>
          <MemoryRouter>
            <UsersPage />
          </MemoryRouter>
        </UserProvider>
      )

      const deleteUsersButton = screen.getByRole('button', {
        name: /Users.Actions.DeleteUserBulk/i,
      })

      expect(deleteUsersButton).toBeDisabled()

      const [firstRow] = screen.getAllByRole('row', { name: /Select row/i })
      const rowSelectCheckbox = getByRole(firstRow, 'checkbox')
      await userEvent.click(rowSelectCheckbox)

      expect(deleteUsersButton).not.toBeDisabled()
    })
  })

  describe('navigation', () => {
    test('Username link navigates correctly', async () => {
      let router = createMemoryRouter(
        [
          {
            path: 'users',
            element: <UsersPage />,
          },
          {
            path: 'users/create',
            element: <UserForm />,
          },
          {
            path: 'users/:userId',
            element: <UserDetails />,
          },
          {
            path: 'users/:userId/edit',
            element: <UserForm />,
          },
        ],
        {
          initialEntries: ['/users'],
        }
      )

      render(
        <UserProvider>
          <RouterProvider router={router} />
        </UserProvider>
      )

      const [firstRow] = screen.getAllByRole('row', { name: /Select row/i })
      const usernameLink = getByRole(firstRow, 'link')
      expect(usernameLink).toBeInTheDocument()
      expect(
        screen.queryByRole('button', { name: /Users.Actions.UpdateUser/i })
      ).not.toBeInTheDocument()

      await userEvent.click(usernameLink)

      expect(
        screen.getByRole('button', { name: /Users.Actions.UpdateUser/i })
      ).toBeInTheDocument()
    })
  })

  describe('CRUD', () => {
    // TODO:
    test('create user', () => {})

    // TODO:
    test('read user', () => {})

    // TODO:
    test('update user', () => {})

    // TODO:
    test('delete user', () => {})
  })
})
