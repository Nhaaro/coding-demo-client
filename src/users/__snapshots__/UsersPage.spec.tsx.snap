// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersPage presentation it can render with not data 1`] = `
<div
  class="MuiDataGrid-virtualScrollerRenderZone css-s1v7zr-MuiDataGrid-virtualScrollerRenderZone"
  role="rowgroup"
  style="transform: translate3d(0, 0px, 0);"
/>
`;

exports[`UsersPage presentation table starts with correct data 1`] = `
[
  <div
    aria-rowindex="2"
    aria-selected="false"
    class="MuiDataGrid-row MuiDataGrid-row--firstVisible"
    data-id="1"
    data-rowindex="0"
    role="row"
    style="max-height: 52px; min-height: 52px; --height: 52px;"
  >
    <div
      class="MuiDataGrid-cellOffsetLeft"
      role="presentation"
      style="width: 0px;"
    />
    <div
      aria-colindex="1"
      aria-colspan="1"
      class="MuiDataGrid-cellCheckbox MuiDataGrid-cell--flex MuiDataGrid-cell MuiDataGrid-cell--textCenter"
      data-colindex="0"
      data-field="__check__"
      role="gridcell"
      style="--width: 50px;"
      tabindex="-1"
    >
      <span
        class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiDataGrid-checkboxInput css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
      >
        <input
          aria-label="Select row"
          class="PrivateSwitchBase-input css-1m9pwf3"
          data-indeterminate="false"
          tabindex="-1"
          type="checkbox"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="CheckBoxOutlineBlankIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </span>
    </div>
    <div
      aria-colindex="2"
      aria-colspan="1"
      class="MuiDataGrid-cell MuiDataGrid-cell--textLeft"
      data-colindex="1"
      data-field="id"
      role="gridcell"
      style="--width: 100px;"
      tabindex="-1"
      title="1"
    >
      1
    </div>
    <div
      aria-colindex="3"
      aria-colspan="1"
      class="MuiDataGrid-cell MuiDataGrid-cell--textLeft"
      data-colindex="2"
      data-field="username"
      role="gridcell"
      style="--width: 180px;"
      tabindex="-1"
    >
      <a
        class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineHover css-11iau0r-MuiTypography-root-MuiLink-root"
        href="/users/1"
      >
        username
      </a>
    </div>
    <div
      aria-colindex="4"
      aria-colspan="1"
      class="MuiDataGrid-cell MuiDataGrid-cell--textLeft"
      data-colindex="3"
      data-field="name"
      role="gridcell"
      style="--width: 250px;"
      tabindex="-1"
      title="User Name"
    >
      User Name
    </div>
    <div
      aria-colindex="5"
      aria-colspan="1"
      class="MuiDataGrid-cell MuiDataGrid-cell--textLeft"
      data-colindex="4"
      data-field="email"
      role="gridcell"
      style="--width: 150px;"
      tabindex="-1"
      title="mail@mail.com"
    >
      mail@mail.com
    </div>
    <div
      aria-colindex="6"
      aria-colspan="1"
      class="MuiDataGrid-cell--flex MuiDataGrid-cell MuiDataGrid-cell--textCenter"
      data-colindex="5"
      data-field="active"
      role="gridcell"
      style="--width: 100px;"
      tabindex="-1"
    >
      <svg
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall MuiDataGrid-booleanCell css-ptiqhd-MuiSvgIcon-root"
        data-testid="CheckIcon"
        data-value="true"
        focusable="false"
        role="img"
        viewBox="0 0 24 24"
      >
        <path
          d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z"
        />
        <title>
          yes
        </title>
      </svg>
    </div>
    <div
      aria-colindex="7"
      aria-colspan="1"
      class="MuiDataGrid-cell MuiDataGrid-cell--textLeft"
      data-colindex="6"
      data-field="createdAt"
      role="gridcell"
      style="--width: 200px;"
      tabindex="-1"
      title="16/07/2024, 05:38:14"
    >
      16/07/2024, 05:38:14
    </div>
    <div
      aria-colindex="8"
      aria-colspan="1"
      class="MuiDataGrid-cell--flex MuiDataGrid-cell MuiDataGrid-cell--textCenter"
      data-colindex="7"
      data-field="actions"
      role="gridcell"
      style="--width: 100px;"
      tabindex="-1"
    >
      <div
        class="MuiDataGrid-actionsCell"
        role="menu"
        tabindex="-1"
      >
        <a
          aria-label="Users.Actions.UpdateUser"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          data-mui-internal-clone-element="true"
          href="/users/1/edit"
          role="menuitem"
          tabindex="-1"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="EditIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </a>
        <button
          aria-label="Users.Actions.DeleteUser"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
          data-mui-internal-clone-element="true"
          role="menuitem"
          tabindex="-1"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
            data-testid="DeleteIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>,
  <div
    aria-rowindex="3"
    aria-selected="false"
    class="MuiDataGrid-row MuiDataGrid-row--lastVisible"
    data-id="12"
    data-rowindex="1"
    role="row"
    style="max-height: 52px; min-height: 52px; --height: 52px;"
  >
    <div
      class="MuiDataGrid-cellOffsetLeft"
      role="presentation"
      style="width: 0px;"
    />
    <div
      aria-colindex="1"
      aria-colspan="1"
      class="MuiDataGrid-cellCheckbox MuiDataGrid-cell--flex MuiDataGrid-cell MuiDataGrid-cell--textCenter"
      data-colindex="0"
      data-field="__check__"
      role="gridcell"
      style="--width: 50px;"
      tabindex="-1"
    >
      <span
        class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiDataGrid-checkboxInput css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
      >
        <input
          aria-label="Select row"
          class="PrivateSwitchBase-input css-1m9pwf3"
          data-indeterminate="false"
          tabindex="-1"
          type="checkbox"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="CheckBoxOutlineBlankIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </span>
    </div>
    <div
      aria-colindex="2"
      aria-colspan="1"
      class="MuiDataGrid-cell MuiDataGrid-cell--textLeft"
      data-colindex="1"
      data-field="id"
      role="gridcell"
      style="--width: 100px;"
      tabindex="-1"
      title="12"
    >
      12
    </div>
    <div
      aria-colindex="3"
      aria-colspan="1"
      class="MuiDataGrid-cell MuiDataGrid-cell--textLeft"
      data-colindex="2"
      data-field="username"
      role="gridcell"
      style="--width: 180px;"
      tabindex="-1"
    >
      <a
        class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineHover css-11iau0r-MuiTypography-root-MuiLink-root"
        href="/users/12"
      >
        otheruser
      </a>
    </div>
    <div
      aria-colindex="4"
      aria-colspan="1"
      class="MuiDataGrid-cell MuiDataGrid-cell--textLeft"
      data-colindex="3"
      data-field="name"
      role="gridcell"
      style="--width: 250px;"
      tabindex="-1"
      title="Another User"
    >
      Another User
    </div>
    <div
      aria-colindex="5"
      aria-colspan="1"
      class="MuiDataGrid-cell MuiDataGrid-cell--textLeft"
      data-colindex="4"
      data-field="email"
      role="gridcell"
      style="--width: 150px;"
      tabindex="-1"
      title="other@mail.com"
    >
      other@mail.com
    </div>
    <div
      aria-colindex="6"
      aria-colspan="1"
      class="MuiDataGrid-cell--flex MuiDataGrid-cell MuiDataGrid-cell--textCenter"
      data-colindex="5"
      data-field="active"
      role="gridcell"
      style="--width: 100px;"
      tabindex="-1"
    >
      <svg
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall MuiDataGrid-booleanCell css-ptiqhd-MuiSvgIcon-root"
        data-testid="CheckIcon"
        data-value="true"
        focusable="false"
        role="img"
        viewBox="0 0 24 24"
      >
        <path
          d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z"
        />
        <title>
          yes
        </title>
      </svg>
    </div>
    <div
      aria-colindex="7"
      aria-colspan="1"
      class="MuiDataGrid-cell MuiDataGrid-cell--textLeft"
      data-colindex="6"
      data-field="createdAt"
      role="gridcell"
      style="--width: 200px;"
      tabindex="-1"
      title="19/07/2024, 03:55:22"
    >
      19/07/2024, 03:55:22
    </div>
    <div
      aria-colindex="8"
      aria-colspan="1"
      class="MuiDataGrid-cell--flex MuiDataGrid-cell MuiDataGrid-cell--textCenter"
      data-colindex="7"
      data-field="actions"
      role="gridcell"
      style="--width: 100px;"
      tabindex="-1"
    >
      <div
        class="MuiDataGrid-actionsCell"
        role="menu"
        tabindex="-1"
      >
        <a
          aria-label="Users.Actions.UpdateUser"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          data-mui-internal-clone-element="true"
          href="/users/12/edit"
          role="menuitem"
          tabindex="-1"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="EditIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </a>
        <button
          aria-label="Users.Actions.DeleteUser"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
          data-mui-internal-clone-element="true"
          role="menuitem"
          tabindex="-1"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
            data-testid="DeleteIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>,
]
`;
